xray/grpc.go

```diff
@@ -0,0 +1,256 @@
+		}
+
+		return resp, err
+	}
+}
+
+func classifyErrorStatus(seg *Segment, err error) {
+	seg.Lock()
+	defer seg.Unlock()
+	grpcStatus, ok := status.FromError(err)
+	if !ok {
+		seg.Fault = true
+		return
+	}
+	switch grpcStatus.Code() {
+	case codes.Canceled, codes.InvalidArgument, codes.NotFound, codes.AlreadyExists, codes.PermissionDenied, codes.Unauthenticated, codes.FailedPrecondition, codes.Aborted, codes.OutOfRange:
+		seg.Error = true
+	case codes.Unknown, codes.DeadlineExceeded, codes.Unimplemented, codes.Internal, codes.Unavailable, codes.DataLoss:
+		seg.Fault = true
+	case codes.ResourceExhausted:
+		seg.Throttle = true
+	}
+}
+
+func clientIPFromGrpcMetadata(md metadata.MD) (string, bool) {
+	if len(md.Get("x-forwarded-for")) != 1 {
+		return "", false
+	}
+	forwardedFor := md.Get("x-forwarded-for")[0]
+	if forwardedFor != "" {
+		return strings.TrimSpace(strings.Split(forwardedFor, ",")[0]), true
+	}
+	return "", false
+}
+
+func recordContentLength(seg *Segment, reply interface{}) {
+	seg.Lock()
+	defer seg.Unlock()
+	if protoMessage, isProtoMessage := reply.(proto.Message); isProtoMessage {
+		seg.GetHTTP().GetResponse().ContentLength = proto.Size(protoMessage)
+	}
+}
+
+func addResponseTraceHeader(ctx context.Context, seg *Segment, incomingTraceHeader *header.Header) error {
+	var respHeader bytes.Buffer
+	respHeader.WriteString("Root=")
+	respHeader.WriteString(seg.TraceID)
+	if incomingTraceHeader.SamplingDecision == header.Requested {
+		respHeader.WriteString(";Sampled=")
+		respHeader.WriteString(strconv.Itoa(btoi(seg.Sampled)))
+	}
+
+	headers := metadata.New(map[string]string{
+		TraceIDHeaderKey: respHeader.String(),
+	})
+	return grpc.SendHeader(ctx, headers)
+}
+
+func inferServiceName(fullMethodName string) string {
+	fullMethodName = fullMethodName[1:]
+	return fullMethodName[:strings.Index(fullMethodName, "/")]
+}
+
+type ServerInterceptorOption interface {
+	apply(option *serverInterceptorOption)
+}
+
+type serverInterceptorOption struct {
+	ctx          context.Context
+	segmentNamer SegmentNamer
+}
+
+func newFuncServerInterceptorOption(f func(option *serverInterceptorOption)) ServerInterceptorOption {
+	return funcServerInterceptorOption{f: f}
+}
+
+type funcServerInterceptorOption struct {
+	f func(option *serverInterceptorOption)
+}
+
+func (f funcServerInterceptorOption) apply(option *serverInterceptorOption) {
+	f.f(option)
+}
+
+// ServerInterceptorWithContext makes the interceptor inherit xray.Config associated with ctx.
+func ServerInterceptorWithContext(ctx context.Context) ServerInterceptorOption {
+	return newFuncServerInterceptorOption(func(option *serverInterceptorOption) {
+		option.ctx = ctx
+	})
+}
+
+// ServerInterceptorWithSegmentNamer makes the interceptor use the segment namer to name the segment.
+func ServerInterceptorWithSegmentNamer(sn SegmentNamer) ServerInterceptorOption {
+	return newFuncServerInterceptorOption(func(option *serverInterceptorOption) {
+		option.segmentNamer = sn
+	})
+}
```

Comment:
Maybe we should rename this into `UnaryServerInterceptorWithSegmentNamer`? Or organize interceptors and options into sub-packages?

We might need to find a proper way to organize those options because interceptors of gRPC have four variants:
1. Unary / Client
2. Unary / Server

In the future, we might implement interceptors for gRPC streaming. Then we will have:
3. Stream / Client
4. Stream / Server

I can imagine, different types of interceptors will have different types of concerns and options.